### DISCUSSION TITLE
Rate limit

### DISCUSSION THREAD

**Discussion by zohouri** [sindresorhus/got#2285] [Q&A] [answered]
Hi,
Is it possible to apply a rate limit to a Got instance? For example, only send 30 requests per minute (if we have more requests, don't cancel them or throw an error, but delay them so that the rate limit is met)

**Comment by sindresorhus** [comment, marked as answer]
There is no built-in rate limiting as there are simply too many variations of rate limiting. To achieve it, you could [make a plugin](https://github.com/sindresorhus/got/blob/b1d61c173a681755ac23afb2f155f08801c1e7e4/documentation/lets-make-a-plugin.md?plain=1#L207).

You can also easily implement it using [`p-queue`](https://github.com/sindresorhus/p-queue):

```javascript
import PQueue from 'p-queue';
import got from 'got';

// Limit to 30 requests per minute
const queue = new PQueue({
	intervalCap: 30,
	interval: 60000, // 60 seconds
	concurrency: 30 // Optional: control concurrent requests
});

// Wrap your got calls in queue.add()
async function makeRequest(url) {
	return queue.add(() => got(url).json());
}

// Use it
const results = await Promise.all([
	makeRequest('https://api.example.com/1'),
	makeRequest('https://api.example.com/2'),
	makeRequest('https://api.example.com/3')
	// ... add as many as you want
]);
```

This will automatically delay excess requests instead of canceling them, which is exacty what you're looking for. The queue handles all the timing logic for you, so you don't have to worry about implementing the rate limiting yourself.
